### ISSUE TITLE
High memory usage & complete openHAB shutdown with many SSE subscriptions

### ISSUE THREAD

**Issue by steilerDev** [openhab/openhab-core#598]
Hi All,

as already mentioned in [an issue on the openHAB-distro repo](https://github.com/openhab/openhab-distro/issues/862) I've been experiencing a high memory and CPU usage after some time of running openHAB, up until the point, where the whole system becomes unresponsive after a couple of hours/days.

I developed a Homebridge plugin that uses a single subscription for every item used in Homebridge through the `http://<openhab-host>/rest/events?topics=smarthome/items/<habItem>/statechanged` interface. Due to the size of my installation I ended up having about 60 of those subscriptions opened. The updates should have been delivered consistently and all subscriptions were kept opened over the time (no stale subscriptions afaik). Nevertheless after changing the model to use a single subscription to `http://<openhab-host>/rest/events?topics=smarthome/items/` and filtering the relevant items on the receiving side I have not experienced any memory or CPU issues, therefore I think there is some unexpected behavior in this area.

I created a dump file using `dev:dump-create` from the console during the time, where the issue was happening, but I am not sure how to interpret the information. [Maybe you can have a look at that](https://www.dropbox.com/s/4z5nv63zjncqgg0/2019-02-18_123012.zip?dl=0)

I will continue looking at the issue and try to create a PoC. Until then maybe you can give your feedback/experience with that.

Thanks,

Frank

**Comment by wborn** [comment]
In the heapdump 527MB was used by the `SseResource` (while 531MB of memory was still available):

![1](https://user-images.githubusercontent.com/12213581/53302605-33e25e80-3860-11e9-92bc-bbaff29a1e5c.png)

--- 

Besides the `SseResource`, there's nothing else consuming significant amounts of memory:

![2](https://user-images.githubusercontent.com/12213581/53302617-5c6a5880-3860-11e9-9dee-9dccada239d0.png)

---

All memory is in the [chunkedOutputs](https://github.com/jersey/jersey/blob/faa809da43538ce31076b50f969b4bd64caa5ac9/core-server/src/main/java/org/glassfish/jersey/server/Broadcaster.java#L64):

![3](https://user-images.githubusercontent.com/12213581/53302926-8b82c900-3864-11e9-9df1-1c3e45498c58.png)

---

The `ChunkedOutputs` grows by calls to [add(..)](https://github.com/jersey/jersey/blob/faa809da43538ce31076b50f969b4bd64caa5ac9/core-server/src/main/java/org/glassfish/jersey/server/Broadcaster.java#L99-L101). Which is called in 
https://github.com/openhab/openhab-core/blob/45858f2dd2a5ba53f713e593a499ecb859b7345a/bundles/org.openhab.core.io.rest.sse/src/main/java/org/eclipse/smarthome/io/rest/sse/SseResource.java#L99-L108

---

Looks like there are 78505 `SseEventOutput`s which is a lot if there should only be 60 subscriptions:

![selection_002](https://user-images.githubusercontent.com/12213581/53303383-faaeec00-3869-11e9-80dd-807c6da9d46f.png)

**Comment by steilerDev** [comment]
I'm not sure what your point is. Do you consider this as expected behavior, or should we invest more time on this?

**Comment by wborn** [comment]
It looks like your installation created 78505 subscriptions instead of 60.

**Comment by maggu2810** [comment]
@steilerDev Thanks for your report and your dump.
@wborn Thanks for your analysis.

I don't think it is expected behaviour that 527MB are consumed by the ``SseResource` for 60 descriptions.

If the number of subscriptions is 78505 it is only about 6.9 kb per subscriptions. Then it should be okay.

So, are there dangling subscriptions that get never cleaned up?
We should have a look at.

**Comment by davidgraeff** [comment]
Because I had my play with SSE in my Paper UI NG design study and I basically loose the SSE connection every 2 minutes (same with Paper UI, as I have found out), and my OHs memory footprint also increases (but gets garbage collected at some point) my question:

Can we switch to websockets for OH3?

**Comment by maggu2810** [comment]
Hm, I assume for openHAB itself it should not make a lot of difference with respect to the code.

But with respect to (reverse) proxies handling WebSockets is IMHO much more complicated.
I also don't see the real "requirements".

As long as we would like to provide REST interfaces (and I assume we don't want to drop them), we only need a unidirectional channel and not a directional.
Do you see any advantage using WebSockets if it is used only for pushing messages from the server to the client?

If the current SSE implementation is buggy, perhaps we should consider to fix them or use another one instead of migrate to another technology.

**Comment by davidgraeff** [comment]
> Do you see any advantage using WebSockets if it is used only for pushing messages from the server to the client?

You are right that we only need a unidirectional channel. The feature of websockets that I'm after is: life-cycle. We know exactly when a session is active and we wouldn't need the current sitemap "register" concept anymore. You just request a websocket connection and in the http header for the request you may authorization and filter parameters. And as long as the websocket session is active we keep sending interesting events to the client.

**Comment by maggu2810** [comment]
The current problem of this issue exists in the "normal" SSE eventing of component of openHAB and not the sitemap one that use a custom registration concept.
For such an registration some ping pong is necessary.

I would assume that a normal SSE connection will detect a closed connection, too. At least if the server does not receive any TCP ack for the sent events.

**Comment by davidgraeff** [comment]
> I would assume that a normal SSE connection will detect a closed connection, too.

I'd assume the same, but that's what didn't work here, I guess. The issue is: It is really easy for a client to re-request /events, whereas with websockets you actively need to create the socket and have some a bit more sophisticated session logic than with SSE. That's why I thought it might help here.
